### ISSUE TITLE
External LM : Performance reduction after adding external LM

### ISSUE THREAD

**Issue by Saravinnai** [espnet/espnet#4209]
Hi All,
 We have trained the conformer model for around 550 hrs of custom Hindi data. For LM also used the same data.
 We tried external LM (https://github.com/espnet/espnet/blob/master/egs2/librispeech/asr1/conf/tuning/train_lm_transformer2.yaml),
RNNLM (https://github.com/espnet/espnet/blob/master/egs2/librispeech/asr1/conf/tuning/train_lm_adam.yaml)) to improve the performance but performance degradation happened after adding External LM.
 Information About the Dataset: Conformer Model Training data: 550hrs of AM data
 Transformer LM Train text: Used Same 550hrs of data for LM Training consisting of 3.3Lacs lines
 **Ex. WER**
 Conformer BPE 5000:
     %WER 32.08 [ 13251 / 41303, 2026 ins, 2596 del, 8629 sub ]
 Conformer Transformer: LM weight 0.6
     %%WER 40.44 [ 16704 / 41303, 1843 ins, 5508 del, 9353 sub ]
 Conformer RNNLM: LM weight 0.6
     %WER 54.99 [ 22711 / 41303, 3089 ins, 7282 del, 12340 sub ]
 Want to know
1. why WER is not improving with external lm?
2. Is it because of Less training data?
3. is it because of the same LM and AM data we are using?
4. Do we have to use different LM configurations for improving WER,?
 Can anyone help me to understand this?
 Thanks
 Regards
 Saranya

**Comment by sw005320** [comment]
If we use the same data for LM training, the performance improvement is marginal since the encoder-decoder or transducer has its own LM-like sub-networks.
So, for me, this result is not so surprising.

However, your LM weight seems to be too large. I recommend you use a small weight, e.g., 0.1.

**Comment by Saravinnai** [comment]
thanks for the reply. Had a few more questions,
1. if we are using external LM, how much additional data is required to get improvement in WER?
2. Usually higher LM weight will give an additional boost to external LM (domain-specific LM), but why do we have to reduce it to less LM weight of 0.1 ?

**Comment by sw005320** [comment]
> 1. if we are using external LM, how much additional data is required to get improvement in WER?

I don't have a good answer for it, but we usually use 10x or 100x text data.
But note that an external LM would not work if the topic is mismatched.

> 2. Usually higher LM weight will give an additional boost to external LM (domain-specific LM), but why do we have to reduce it to less LM weight of 0.1 ?

This is the case that the topic is really matched and also it is trained with large corpora (like WSJ). This does not happen in most ASR scenarios.

**Comment by Saravinnai** [comment]
okay. Thanks. Few more questions,
1. If a topic mismatch is there between the trained conformer model and external LM, we will not see any improvement in WER?
2. Supposed we have trained conformer model with general data, and our target use case is call center ASR. With call center data, we are getting around 33 % WER. We want to reduce the WER. Is adding only call center conversation text in terms of external LM will be helpful? Or do you suggest any other pointer?
3. For our custom 550 hrs of Hindi data, in Kaldi, we got 31 % WER, and with the conformer model with BPE of 5000 WER increased to 32%. So do you suggest any method to improve the WER of our model?

**Comment by sw005320** [comment]
> 1. If a topic mismatch is there between the trained conformer model and external LM, we will not see any improvement in WER?

Yes

> 2. Supposed we have trained conformer model with general data, and our target use case is call center ASR. With call center data, we are getting around 33 % WER. We want to reduce the WER. Is adding only call center conversation text in terms of external LM will be helpful? Or do you suggest any other pointer?

Yes, helpful, but the LM gain is not very large in general. 

> 3. For our custom 550 hrs of Hindi data, in Kaldi, we got 31 % WER, and with the conformer model with BPE of 5000 WER increased to 32%. So do you suggest any method to improve the WER of our model?

You can tune the model architecture and optimization hyperparameters first.

**Comment by Saravinnai** [comment]
Currently, we are using this configuration: 
https://github.com/espnet/espnet/blob/6f429608bedc8d63d5c4cc401c7f5ed738da5dee/egs2/librispeech/asr1/conf/tuning/train_asr_conformer6_n_fft512_hop_length256.yaml
We tried some hyper-parameters tuning like changing learning rate and BPE. Our observations are:
a. With a learning rate (LR) of 0.0015, we didn’t get a good result, so we changed to LR 0.001 and warmup_steps: 40000.  We kept the default BPE=5000, With this configuration we got
	%WER 32.08 [ 13251 / 41303, 2026 ins, 2596 del, 8629 sub ]
b. Additionally, we tried to change BPE size to 1000 and 10000
	With BPE 1000: %WER 32.77 [ 13534 / 41303, 1944 ins, 3038 del, 8552 sub ]
	With BPE 10000: the model didn’t train properly due to the grad nan issue
So can you suggest
Which hyperparameters we can change for our custom 550 hrs of Hindi data for improving WER?
Do you suggest any other model architecture apart from 
https://github.com/espnet/espnet/blob/6f429608bedc8d63d5c4cc401c7f5ed738da5dee/egs2/librispeech/asr1/conf/tuning/train_asr_conformer6_n_fft512_hop_length256.yaml
If a topic mismatch is there between our trained model and testing data. Do you suggest any method to improve WER like ngram or external LM etc? Usually, in the Kaldi model, we see n-gram model is very effective for improving WER.

**Comment by sw005320** [comment]
> thanks. Currently, we are using this configuration: **https://github.com/espnet/espnet/blob/6f429608bedc8d63d5c4cc401c7f5ed738da5dee/egs2/librispeech/asr1/conf/tuning/train_asr_conformer6_n_fft512_hop_length256.yaml**
> 
> We tried some hyper-parameters tuning like changing learning rate and BPE. Our observations are:
> 
> a. With a learning rate (LR) of 0.0015, we didn’t get a good result, so we changed to LR 0.001 and warmup_steps: 40000.
> 
> We kept the default BPE=5000, With this configuration we got
> 
> `%WER 32.08 [ 13251 / 41303, 2026 ins, 2596 del, 8629 sub ]`
> 
> b. Additionally, we tried to change BPE size to 1000 and 10000
> 
> ```
> With BPE 1000: %WER 32.77 [ 13534 / 41303, 1944 ins, 3038 del, 8552 sub ]
> With BPE 10000: the model didn’t train properly due to the grad nan issue
> ```

The BPE size is too large.
We usually use around thousands.

> So can you suggest
> 
> 1. Which hyperparameters we can change for our custom 550 hrs of Hindi data for improving WER?

I recommend you check our configs https://github.com/espnet/espnet/tree/master/egs2/librispeech/asr1/conf/tuning and the corresponding documents https://github.com/espnet/espnet/blob/master/egs2/librispeech/asr1/README.md to see which parameter we tuned to get the current librispeech numbers.
We tuned the number of layers, hopsize, learning rate, layerdrop, etc.

> 2. Do you suggest any other model architecture apart from
>    **https://github.com/espnet/espnet/blob/6f429608bedc8d63d5c4cc401c7f5ed738da5dee/egs2/librispeech/asr1/conf/tuning/train_asr_conformer6_n_fft512_hop_length256.yaml**

We usually do not change the architecture.
Conformer is working well.

> 3. If a topic mismatch is there between our trained model and testing data. Do you suggest any method to improve WER like ngram or external LM etc? Usually, in the Kaldi model, we see n-gram model is very effective for improving WER.

Do you mean that you have a large amount of text data for the testing data?
Then, LM shallow fusion would work.
